### ISSUE TITLE
Unexpected Reconnect event with infinite loop of 401 error and connectInterval at 0

### ISSUE THREAD

**Issue by SimonBriche** [ttezel/twit#253]
Hi,

I have updated twit to the latest version and I'm experiencing now unexpected reconnect events with 401 error, which breaks tweets listening. Here is an extract of my logs that shows this behavior :

```
» 14 Mar 2016 10:41:08.904 2016-03-14 09:42:16.464454+00:00 app listener.1 - - twit connected
» 14 Mar 2016 10:41:08.985 2016-03-14 09:42:16.464531+00:00 app listener.1 - - twit connected status code : 200
» 15 Mar 2016 11:18:02.097 2016-03-15 10:19:13.794662+00:00 app listener.1 - - twit connected
» 15 Mar 2016 11:18:02.177 2016-03-15 10:19:13.795084+00:00 app listener.1 - - twit connected status code : 200
» 15 Mar 2016 17:16:59.092 2016-03-15 16:18:11.875857+00:00 app listener.1 - - twit reconnect
» 15 Mar 2016 17:16:59.170 2016-03-15 16:18:11.875885+00:00 app listener.1 - - twit response status code : 200
» 15 Mar 2016 17:16:59.170 2016-03-15 16:18:11.875932+00:00 app listener.1 - - twit connectInterval : 0 
» 15 Mar 2016 17:18:29.678 2016-03-15 16:19:42.380796+00:00 app listener.1 - - twit reconnect
» 15 Mar 2016 17:18:29.759 2016-03-15 16:19:42.380899+00:00 app listener.1 - - twit response status code : 401
» 15 Mar 2016 17:18:29.759 2016-03-15 16:19:42.380941+00:00 app listener.1 - - twit connectInterval : 0
» 15 Mar 2016 17:20:00.152 2016-03-15 16:21:12.880327+00:00 app listener.1 - - twit reconnect
» 15 Mar 2016 17:20:00.228 2016-03-15 16:21:12.880906+00:00 app listener.1 - - twit response status code : 401
» 15 Mar 2016 17:20:00.228 2016-03-15 16:21:12.881341+00:00 app listener.1 - - twit connectInterval : 0
» 15 Mar 2016 17:21:30.674 2016-03-15 16:22:43.390669+00:00 app listener.1 - - twit reconnect
» 15 Mar 2016 17:21:30.749 2016-03-15 16:22:43.390755+00:00 app listener.1 - - twit response status code : 401
» 15 Mar 2016 17:21:30.749 2016-03-15 16:22:43.390794+00:00 app listener.1 - - twit connectInterval : 0
» 15 Mar 2016 17:23:01.168 2016-03-15 16:24:13.882719+00:00 app listener.1 - - twit reconnect
» 15 Mar 2016 17:23:01.249 2016-03-15 16:24:13.882754+00:00 app listener.1 - - twit response status code : 401
» 15 Mar 2016 17:23:01.249 2016-03-15 16:24:13.882773+00:00 app listener.1 - - twit connectInterval : 0
» 15 Mar 2016 17:24:31.655 2016-03-15 16:25:44.370671+00:00 app listener.1 - - twit reconnect
» 15 Mar 2016 17:24:31.741 2016-03-15 16:25:44.370808+00:00 app listener.1 - - twit response status code : 401
```

As you can see, there is a 200 'connected' event dispatched every few hours, or at each server cycling. So far so good. And then a 200 'reconnect' event is dispatched "out of the blue" with a reconnectInterval at 0. 2 minutes later, a 401 'reconnect' event is dispatched with a reconnectInterval at 0 and here we go for an infinite loop of 401 'reconnect' events every few minutes (from 1 to 3). If I reboot my server, the connection is successful again, so it's not a credentials issue.

I'm experiencing this behavior with twit@2.2.2 and twit@2.2.3, but not with twit@2.1.1. I have an older application which use this version, hosted on Heroku as well, with the same codebase (at least for twit integration). Here is the extract logs with twit@2.1.1 :

```
» 15 Mar 2016 10:56:54.784 2016-03-15 09:58:05.336134+00:00 app listener.1 - - twit connected
» 15 Mar 2016 10:56:54.867 2016-03-15 09:58:05.336175+00:00 app listener.1 - - twit connected status code : 200
» 15 Mar 2016 17:11:01.003 2016-03-15 16:12:13.710047+00:00 app listener.1 - - twit reconnect
» 15 Mar 2016 17:11:01.085 2016-03-15 16:12:13.710419+00:00 app listener.1 - - twit response status code : 200
» 15 Mar 2016 17:11:01.085 2016-03-15 16:12:13.710543+00:00 app listener.1 - - twit connectInterval : 0
» 15 Mar 2016 17:11:01.645 2016-03-15 16:12:14.308661+00:00 app listener.1 - - twit connected
» 15 Mar 2016 17:11:01.725 2016-03-15 16:12:14.308773+00:00 app listener.1 - - twit connected status code : 200
» 16 Mar 2016 00:59:46.618 2016-03-16 00:01:00.620311+00:00 app listener.1 - - twit connected
» 16 Mar 2016 00:59:46.695 2016-03-16 00:01:00.620352+00:00 app listener.1 - - twit connected status code : 200
```

There is the unexpected 'reconnect' event as well, but immediately followed by a 200 'connected' event, that lacks in twit@2.2.3, and the 401 'reconnect' events aren't dispatched.

So far, this behavior happens randomly every few days (from 1 to 5), thus very difficult to track down. I've noticed that you have changed the timestamp of the OAuth'ed HTTP requests in 2.2.0, and fix the reconnect race in 2.2.3. Have you experienced weird reconnect behaviours as well ?

Technical details :
- My application is hosted on Heroku/Europe (I checked the timestamp of the server and it seems to be correct)
- Here is my twit initialization code that produced the logs :

```
stream = T.stream('statuses/filter', { track: ENV.TWITTER_TRACK});

stream.on('tweet', function (tweet) {
  console.log("tweet");
  checkTweet(tweet, callback);
});

stream.on('connected', function (response) {
  console.log('twit connected');
  if(response) console.log('twit connected status code : '+response.statusCode);
});

stream.on('disconnect', function (disconnectMessage) {
  console.log('twit disconnect : '+disconnectMessage);
});

stream.on('reconnect', function (request, response, connectInterval) {
  console.log('twit reconnect');
  if(response) console.log('twit response status code : '+response.statusCode);
  console.log('twit connectInterval : '+connectInterval);
});

stream.on('limit', function (limitMessage) {
  console.log('twit limit : ');
  console.log(limitMessage);
});

stream.on('warning', function (warning) {
  console.log('twit warning : ');
  console.log(warning);
});

stream.on('error', function (err) {
  console.log('twit error : ');
  console.log(err);
});

stream.on('message', function (msg) {
  console.log('twit msg : ');
  console.log(msg);
});
```

There's no 'warning', 'error' or even 'message' events dispatched.

Thanks,
Simon

UPDATE :
I have the same 401 reconnect behavior with an other Twitter application/Heroku application/Twitter accessTokens for stream connection. So it seems that it's not related to a specific application or Twitter user's credentials.

**Comment by andreapozzetti** [comment]
+1

**Comment by MeanwhileMedia** [comment]
+1

**Comment by MeanwhileMedia** [comment]
Each time these reconnect attempts occur, a "parser-error" event is emitted immediately after with the following error message:

`Error: Gzip error: incorrect header check
at Object.exports.makeTwitError (/.../node_modules/twit/lib/helpers.js:74:13)
at Gunzip.<anonymous> (/.../node_modules/twit/lib/streaming-api-connection.js:112:31)
at emitOne (events.js:95:20)
at Gunzip.emit (events.js:182:7)
at Zlib._handle.onerror (zlib.js:366:10)`

**Comment by venkatperi** [comment]
+1

**Comment by SimonBriche** [comment]
Like MeanwhileMedia said, it seems that this error is related with Gzip support, and the problem appeared with the 2.2.0 release. I had almost 15 days in a row without error, and it poped again yesterday. 

See below the logs with twit@2.2.0 :

````
» 22 Mar 2016 20:01:02.497 2016-03-22 19:01:01.835486+00:00 app listener.1 - - twit reconnect
» 22 Mar 2016 20:01:02.578 2016-03-22 19:01:01.835519+00:00 app listener.1 - - twit response status code : 401
» 22 Mar 2016 20:01:02.578 2016-03-22 19:01:01.835558+00:00 app listener.1 - - twit connectInterval : 0
» 22 Mar 2016 20:04:03.975 2016-03-22 19:04:03.266018+00:00 app listener.1 - - twit error :
» 22 Mar 2016 20:04:04.051 2016-03-22 19:04:03.266244+00:00 app listener.1 - - { [Error: Gzip error: incorrect header check]
» 22 Mar 2016 20:04:04.051 2016-03-22 19:04:03.266246+00:00 app listener.1 - - message: 'Gzip error: incorrect header check',
» 22 Mar 2016 20:04:04.051 2016-03-22 19:04:03.266246+00:00 app listener.1 - - code: null,
» 22 Mar 2016 20:04:04.051 2016-03-22 19:04:03.266247+00:00 app listener.1 - - allErrors: [],
» 22 Mar 2016 20:04:04.051 2016-03-22 19:04:03.266249+00:00 app listener.1 - - twitterReply: '<html>\\n<head>\\n<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>\\n<title>Error 401 Unauthorized</title>\n</head>\n<body>\n<h2>HTTP ERROR: 401</h2>\n<p>Problem accessing \'/1.1/statuses/filter.json?track=ouzetsebastien1%20loofprint\'. Reason:\n<pre> Unauthorized</pre>\n</body>\n</html>\n',
» 22 Mar 2016 20:04:04.051 2016-03-22 19:04:03.266249+00:00 app listener.1 - - statusCode: 401 }
» 22 Mar 2016 20:07:04.344 2016-03-22 19:07:03.757745+00:00 app listener.1 - - twit reconnect
» 22 Mar 2016 20:07:04.425 2016-03-22 19:07:03.757778+00:00 app listener.1 - - twit response status code : 401
» 22 Mar 2016 20:07:04.425 2016-03-22 19:07:03.757819+00:00 app listener.1 - - twit connectInterval : 0
» 22 Mar 2016 20:07:04.942 2016-03-22 19:07:04.236789+00:00 app listener.1 - - twit error :
» 22 Mar 2016 20:08:34.980 2016-03-22 19:08:34.236024+00:00 app listener.1 - - twit reconnect
» 22 Mar 2016 20:08:35.061 2016-03-22 19:08:34.236066+00:00 app listener.1 - - twit response status code : 401
» 22 Mar 2016 20:08:35.061 2016-03-22 19:08:34.236079+00:00 app listener.1 - - twit connectInterval : 0
» 22 Mar 2016 20:08:35.294 2016-03-22 19:08:34.708145+00:00 app listener.1 - - twit error :
» 22 Mar 2016 20:08:35.375 2016-03-22 19:08:34.708293+00:00 app listener.1 - - { [Error: Gzip error: incorrect header check]
» 22 Mar 2016 20:08:35.375 2016-03-22 19:08:34.708300+00:00 app listener.1 - - message: 'Gzip error: incorrect header check',
» 22 Mar 2016 20:08:35.375 2016-03-22 19:08:34.708301+00:00 app listener.1 - - code: null,
» 22 Mar 2016 20:08:35.375 2016-03-22 19:08:34.708303+00:00 app listener.1 - - allErrors: [],
» 22 Mar 2016 20:08:35.375 2016-03-22 19:08:34.708304+00:00 app listener.1 - - twitterReply: '<html>\\n<head>\\n<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>\\n<title>Error 401 Unauthorized</title>\n</head>\n<body>\n<h2>HTTP ERROR: 401</h2>\n<p>Problem accessing \'/1.1/statuses/filter.json?track=ouzetsebastien1%20loofprint\'. Reason:\n<pre> Unauthorized</pre>\n</body>\n</html>\n',
» 22 Mar 2016 20:08:35.375 2016-03-22 19:08:34.708304+00:00 app listener.1 - - statusCode: 401 }
```

And with twit@2.2.3 :

```
» 19 Apr 2016 15:43:22.567 113 <190>1 2016-04-19T13:43:21.887138+00:00 app listener.1 - - twit reconnect
» 19 Apr 2016 15:43:22.646 130 <190>1 2016-04-19T13:43:21.887331+00:00 app listener.1 - - twit response status code : 200 Context
» 19 Apr 2016 15:43:22.646 123 <190>1 2016-04-19T13:43:21.887377+00:00 app listener.1 - - twit connectInterval : 0
» 19 Apr 2016 15:43:23.164 119 <190>1 2016-04-19T13:43:22.468560+00:00 app listener.1 - - twit parser-error :
» 19 Apr 2016 15:43:23.245 144 <190>1 2016-04-19T13:43:22.471054+00:00 app listener.1 - - { [Error: Gzip error: incorrect header check]
» 19 Apr 2016 15:43:23.245 147 <190>1 2016-04-19T13:43:22.471059+00:00 app listener.1 - - message: 'Gzip error: incorrect header check',
» 19 Apr 2016 15:43:23.245 112 <190>1 2016-04-19T13:43:22.471061+00:00 app listener.1 - - code: null,
» 19 Apr 2016 15:43:23.245 115 <190>1 2016-04-19T13:43:22.471061+00:00 app listener.1 - - allErrors: [],
» 19 Apr 2016 15:43:23.245 431 <190>1 2016-04-19T13:43:22.471063+00:00 app listener.1 - - twitterReply: '<html>\\n<head>\\n<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>\\n<title>Error 401 Unauthorized</title>\n</head>\n<body>\n<h2>HTTP ERROR: 401</h2>\n<p>Problem accessing \'/1.1/statuses/filter.json?track=ouzetsebastien1%20loofprint\'. Reason:\n<pre> Unauthorized</pre>\n</body>\n</html>\n',
» 19 Apr 2016 15:43:23.245 118 <190>1 2016-04-19T13:43:22.471063+00:00 app listener.1 - - statusCode: 401 }
» 19 Apr 2016 15:44:53.139 113 <190>1 2016-04-19T13:44:52.462435+00:00 app listener.1 - - twit reconnect
» 19 Apr 2016 15:44:53.220 130 <190>1 2016-04-19T13:44:52.462463+00:00 app listener.1 - - twit response status code : 401
» 19 Apr 2016 15:44:53.220 123 <190>1 2016-04-19T13:44:52.462492+00:00 app listener.1 - - twit connectInterval : 0
» 19 Apr 2016 15:44:53.539 119 <190>1 2016-04-19T13:44:52.951732+00:00 app listener.1 - - twit parser-error :
» 19 Apr 2016 15:44:53.621 144 <190>1 2016-04-19T13:44:52.952093+00:00 app listener.1 - - { [Error: Gzip error: incorrect header check]
» 19 Apr 2016 15:44:53.621 147 <190>1 2016-04-19T13:44:52.952098+00:00 app listener.1 - - message: 'Gzip error: incorrect header check',
» 19 Apr 2016 15:44:53.621 112 <190>1 2016-04-19T13:44:52.952099+00:00 app listener.1 - - code: null,
» 19 Apr 2016 15:44:53.621 115 <190>1 2016-04-19T13:44:52.952100+00:00 app listener.1 - - allErrors: [],
» 19 Apr 2016 15:44:53.621 431 <190>1 2016-04-19T13:44:52.952102+00:00 app listener.1 - - twitterReply: '<html>\\n<head>\\n<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>\\n<title>Error 401 Unauthorized</title>\n</head>\n<body>\n<h2>HTTP ERROR: 401</h2>\n<p>Problem accessing \'/1.1/statuses/filter.json?track=ouzetsebastien1%20loofprint\'. Reason:\n<pre> Unauthorized</pre>\n</body>\n</html>\n',
» 19 Apr 2016 15:44:53.621 118 <190>1 2016-04-19T13:44:52.952116+00:00 app listener.1 - - statusCode: 401 }
» 19 Apr 2016 15:46:23.663 113 <190>1 2016-04-19T13:46:22.953521+00:00 app listener.1 - - twit reconnect
» 19 Apr 2016 15:46:23.749 130 <190>1 2016-04-19T13:46:22.953539+00:00 app listener.1 - - twit response status code : 401
» 19 Apr 2016 15:46:23.749 123 <190>1 2016-04-19T13:46:22.953577+00:00 app listener.1 - - twit connectInterval : 0
» 19 Apr 2016 15:46:24.155 119 <190>1 2016-04-19T13:46:23.459322+00:00 app listener.1 - - twit parser-error :
» 19 Apr 2016 15:46:24.241 144 <190>1 2016-04-19T13:46:23.459813+00:00 app listener.1 - - { [Error: Gzip error: incorrect header check]
» 19 Apr 2016 15:46:24.241 147 <190>1 2016-04-19T13:46:23.459822+00:00 app listener.1 - - message: 'Gzip error: incorrect header check',
» 19 Apr 2016 15:46:24.241 112 <190>1 2016-04-19T13:46:23.459823+00:00 app listener.1 - - code: null,
» 19 Apr 2016 15:46:24.241 115 <190>1 2016-04-19T13:46:23.459824+00:00 app listener.1 - - allErrors: [],
» 19 Apr 2016 15:46:24.241 431 <190>1 2016-04-19T13:46:23.459830+00:00 app listener.1 - - twitterReply: '<html>\\n<head>\\n<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>\\n<title>Error 401 Unauthorized</title>\n</head>\n<body>\n<h2>HTTP ERROR: 401</h2>\n<p>Problem accessing \'/1.1/statuses/filter.json?track=ouzetsebastien1%20loofprint\'. Reason:\n<pre> Unauthorized</pre>\n</body>\n</html>\n',
» 19 Apr 2016 15:46:24.241 118 <190>1 2016-04-19T13:46:23.459831+00:00 app listener.1 - - statusCode: 401 }
````

We can clearly see that's a parsing error (thanks to the new 'parser-error' event) and related with the Gzip header. So, the error process is :
- a random 200 'reconnect' event with a connectInterval at 0
- SOMETIMES immediately followed by a 'parser-error' event related to Gzip header
- infinite loop of 401 'reconnect' events followed by a 'parser-error' event related to Gzip header

Any clues to fix that ?

Thanks a lot.

**Comment by ttezel** [comment]
Hey all, this has been fixed in twit v2.2.4. Run `npm update twit` to get it! Thanks!

**Comment by SimonBriche** [comment]
Great news ! Thanks for this update :)
